internal/pkg/poster/poster.go

```diff
@@ -70,148 +81,146 @@ func (p *Poster) StopPoSting() {
 }
 
 func (p *Poster) startPoStIfNeeded(ctx context.Context, newHead block.TipSet) error {
-	// TODO: Window PoSt (#4010)
-	//p.postMutex.Lock()
-	//defer p.postMutex.Unlock()
-	//
-	//if p.postCancel != nil {
-	//	// already posting
-	//	return nil
-	//}
-	//
-	//tipsetHeight, err := newHead.Height()
-	//if err != nil {
-	//	return err
-	//}
-	//
-	//root, err := p.chain.GetTipSetStateRoot(ctx, newHead.Key())
-	//if err != nil {
-	//	return err
-	//}
-	//
-	//stateView := p.stateViewer.StateView(root)
-	//provingPeriodStart, _, _, err := stateView.MinerProvingPeriod(ctx, p.minerAddr)
-	//if err != nil {
-	//	return nil
-	//}
-	//
-	//if provingPeriodStart > tipsetHeight {
-	//	// it's not time to PoSt
-	//	return nil
-	//}
-	//
-	//ctx, p.postCancel = context.WithCancel(ctx)
-	//go p.doPoSt(ctx, stateView, provingPeriodStart, newHead.Key())
-	//
+	p.postMutex.Lock()
+	defer p.postMutex.Unlock()
+
+	if p.postCancel != nil {
+		// already posting
+		return nil
+	}
+
+	tipsetHeight, err := newHead.Height()
+	if err != nil {
+		return err
+	}
+
+	root, err := p.chain.GetTipSetStateRoot(ctx, newHead.Key())
+	if err != nil {
+		return err
+	}
+
+	stateView := p.stateViewer.StateView(root)
+	index, _, _, challenge, err := stateView.MinerDeadlineInfo(ctx, p.minerAddr, tipsetHeight)
+	if err != nil {
+		return nil
+	}
+
+	// check that we haven't already seen this deadline
+	if int64(index) == p.poStDeadlineIndex {
+		return nil
+	}
+	p.poStDeadlineIndex = int64(index)
```

Comment:
I infer that this logic at the moment won't handle a re-org that causes a different challenge seed for the post. It'll compute the (now-incorrect) proof, submit it, but not notice that it should retry. 

Ok for now if so, but could you please file an issue outlining the known shortcomings?

Reply:
I added some logic to restart the PoSt if the chain randomness changes within the same deadline (implying a reorg). This plus the confidence interval, should make reorg issues unlikely.